Userland/Libraries/LibWeb/Painting/BorderPainting.cpp

```diff
@@ -150,174 +133,167 @@ void paint_border(PaintContext& context, BorderEdge edge, Gfx::FloatRect const&
-            // top-left
-            center.set_x(radius);
-            center.set_y(radius);
-            offset.set_y(1);
-        } else if (from.x() < to.x() && from.y() < to.y()) {
-            // top-right
-            circle_position.set_x(from.x());
-            center.set_y(radius);
-            offset.set_x(-1);
-            offset.set_y(1);
-        } else if (from.x() > to.x() && from.y() < to.y()) {
-            // bottom-right
-            circle_position.set_x(to.x());
-            circle_position.set_y(from.y());
-            offset.set_x(-1);
-        } else if (from.x() > to.x() && from.y() > to.y()) {
-            // bottom-left
-            circle_position.set_y(to.y());
-            center.set_x(radius);
+    // Avoid overlapping pixels on the edges.
+    if (!top_left)
+        top_border_rect.shrink(0, 0, 0, left_border_rect.width());
+    if (!top_right)
+        top_border_rect.shrink(0, right_border_rect.width(), 0, 0);
+    if (!bottom_left)
+        bottom_border_rect.shrink(0, 0, 0, left_border_rect.width());
+    if (!bottom_right)
+        bottom_border_rect.shrink(0, right_border_rect.width(), 0, 0);
+
+    auto border_color_no_alpha = borders_data.top.color;
+    border_color_no_alpha.set_alpha(255);
+
+    // Paint the strait line part of the border:
+    Painting::paint_border(context, Painting::BorderEdge::Top, top_border_rect, borders_data);
+    Painting::paint_border(context, Painting::BorderEdge::Right, right_border_rect, borders_data);
+    Painting::paint_border(context, Painting::BorderEdge::Bottom, bottom_border_rect, borders_data);
+    Painting::paint_border(context, Painting::BorderEdge::Left, left_border_rect, borders_data);
+
+    // Cache the smallest possible bitmap to render just the corners for the border.
+    auto expand_width = abs(int_width(borders_data.left.width) - int_width(borders_data.right.width));
+    auto expand_height = abs(int_width(borders_data.top.width) - int_width(borders_data.bottom.width));
+    Gfx::IntRect corner_mask_rect {
+        0, 0,
+        max(
+            top_left.horizontal_radius + top_right.horizontal_radius + expand_width,
+            bottom_left.horizontal_radius + bottom_right.horizontal_radius + expand_height),
+        max(
+            top_left.vertical_radius + bottom_left.vertical_radius + expand_width,
+            top_right.vertical_radius + bottom_right.vertical_radius + expand_height)
+    };
+    auto allocate_mask_bitmap = [&]() -> RefPtr<Gfx::Bitmap> {
+        auto bitmap = Gfx::Bitmap::try_create(Gfx::BitmapFormat::BGRA8888, corner_mask_rect.size());
+        if (!bitmap.is_error())
+            return bitmap.release_value();
+        return nullptr;
+    };
+    static thread_local auto corner_bitmap = allocate_mask_bitmap();
+
+    // Only reallocate the corner bitmap is the existing one is too small.
+    // (should mean no more allocations after the first paint -- amortised zero allocations :^))
+    Gfx::Painter painter = ({
```

Comment:
Can we use an IIFE lambda for this instead of relying on compiler extensions more than necessary?

Reply:
Not that I know of, since I want to bail out the function if it fails for some reason (but I also don't really want to propagate this or keep it wrapped in a ErrorOr/Optional -- but maybe I'm missing some obvious)

Reply:
I also don't really want to crash in the rare case that you can't (since the page is still perfectly viewable without the corners)

Reply:
ah, it has a return, alright then